codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs

```diff
@@ -1629,11 +1641,176 @@ ensureNameLookupTables = do
+    |]
+  execute_
+    [here|
+      CREATE TABLE IF NOT EXISTS scoped_type_name_lookup (
+        root_branch_hash_id INTEGER NOT NULL REFERENCES hash(id),
+        -- The name of the term: E.g. List.base
+        reversed_name TEXT NOT NULL,
+        -- The last name segment of the name. This is used when looking up names for
+        -- suffixification when building PPEs.
+        -- E.g. for the name 'base.List.map' this would be 'map'
+        last_name_segment TEXT NOT NULL,
+        -- The namespace containing this definition, not reversed, with a trailing '.'
+        -- The trailing '.' simplifies GLOB queries, so that 'base.*' matches both things in
+        -- 'base' and 'base.List', but not 'base1', which allows us to avoid an OR in our where
+        -- clauses which in turn helps the sqlite query planner use indexes more effectively.
+        --
+        -- example value: 'base.List.'
+        namespace TEXT NOT NULL,
+        reference_builtin TEXT NULL,
+        reference_component_hash INTEGER NULL,
+        reference_component_index INTEGER NULL,
+        PRIMARY KEY (reversed_name, reference_builtin, reference_component_hash, reference_component_index)
+      );
+    |]
+
+  -- This index allows finding all names we need to consider within a given namespace for
+  -- suffixification of a name.
+  -- It may seem strange to use last_name_segment rather than a suffix search over reversed_name name here;
+  -- but SQLite will only optimize for a single prefix-glob at once, so we can't glob search
+  -- over both namespace and reversed_name, but we can EXACT match on last_name_segment and
+  -- then glob search on the namespace prefix, and have SQLite do the final glob search on
+  -- reversed_name over rows with a matching last segment without using an index and should be plenty fast.
+  execute_
+    [here|
+      CREATE INDEX IF NOT EXISTS scoped_type_names_by_namespace_and_last_name_segment ON scoped_type_name_lookup(root_branch_hash_id, last_name_segment, namespace)
+    |]
+
+  -- This index allows us to find all names with a given ref within a specific namespace.
+  execute_
+    [here|
+      CREATE INDEX IF NOT EXISTS scoped_type_name_by_reference_lookup ON scoped_type_name_lookup(root_branch_hash_id, reference_builtin, reference_component_hash, reference_component_index, namespace)
+    |]
+
+  -- Allows fetching ALL names within a specific namespace prefix. We currently use this to
+  -- pretty-print on share, but will be replaced with a more precise set of queries soon.
+  execute_
+    [here|
+      CREATE INDEX IF NOT EXISTS scoped_type_names_by_namespace ON scoped_type_name_lookup(root_branch_hash_id, namespace)
+    |]
+
+-- | Copies existing name lookup rows but replaces their branch hash id;
+-- This is a low-level operation used as part of deriving a new name lookup index
+-- from an existing one as performantly as possible.
+copyScopedNameLookup :: BranchHashId -> BranchHashId -> Transaction ()
```

Comment:
Unfortunately, it seems this isn't as fast as I'd like, since it still needs to re-index the copied rows; but I'm not sure we can possibly make this any faster unless we somehow index every sub-branch independently, but I'm not exactly sure what that would look like or how it would work.

I test this query on a term lookup with 395147 terms in it and it took ~3s.

The nice thing is that now that these indexes are immutable, it'll be much easier and safer to move index generation to background jobs if we want to.